cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkTransactional.java

```diff
@@ -232,6 +260,17 @@ private TransactionalDatasetContext(Transaction transaction,
       this.completion = asyncCommit ? new CountDownLatch(1) : null;
     }
 
+    void useSyncCommit() {
+      // This shouldn't happen as the executeLong will check for isJobStarted before calling this method
```

Comment:
It shouldn't. This method will only get called when there is a implicit transaction in the current thread and then the user declare to use an explicit transaction. E.g.

```scala
val rdd: RDD[(String, Int)] = sc.fromDataset("somedataset")

// The following line will will open an implicit transaction, because Spark will call "rdd.getPartitions()", 
// which we need to get a Dataset instance in order to call "getSplits", however, no Spark job is submitted yet,
// hence "commitOnJobEnded" will be true and "isJobStarted" will be false
val countByKeyRDD = rdd.reduceByKey(_ + _)   

// Next the user starts an explicit transaction. This is when the useSyncCommit() gets invoked. This
// is essentially converting an implicit transaction into an explicit one.
Transaction {   
  val map = countByKeyRDD.collectAsMap()   // Spark will submit a job with an explicit transaction, hence won't commit when job end
}   // Commit happens here
```

If the user execute the above code in multiple threads, there will be multiple implicit tx created and get converted into multiple explicit tx and they won't affect each other.

If the user program uses an explicit transaction covering execution in multiple threads, this method won't get called.